commit 84b3dff9b0680a7c2e5ce7f1066fedc9ef036560 (refs/changes/92/4592/3)
Author: Patrick Mooney <pmooney@pfmooney.com>
Date:   2018-07-30T14:52:23+00:00 (1 year, 2 months ago)
    
    OS-7096 installctx needs kpreempt_disable protection
    Reviewed by: John Levon <john.levon@joyent.com>
    Reviewed by: Jerry Jelinek <jerry.jelinek@joyent.com>
    Reviewed by: Robert Mustacchi <rm@joyent.com>
    Approved by: Robert Mustacchi <rm@joyent.com>

diff --git a/usr/src/uts/common/disp/thread.c b/usr/src/uts/common/disp/thread.c
index 2cd11a0bce..af000bf4f1 100644
--- a/usr/src/uts/common/disp/thread.c
+++ b/usr/src/uts/common/disp/thread.c
@@ -1068,7 +1068,12 @@ installctx(
 	 * form a loop of the ctxops in newest-to-oldest order.  The 'prev'
 	 * pointers form a loop in the reverse direction, where t_ctx->prev is
 	 * the oldest entry associated with the thread.
+	 *
+	 * The protection of kpreempt_disable is required to safely perform the
+	 * list insertion, since there are inconsistent states between some of
+	 * the pointer assignments.
 	 */
+	kpreempt_disable();
 	if (t->t_ctx == NULL) {
 		ctx->next = ctx;
 		ctx->prev = ctx;
@@ -1081,6 +1086,7 @@ installctx(
 		tail->next = ctx;
 	}
 	t->t_ctx = ctx;
+	kpreempt_enable();
 }
 
 /*
